### PR TITLE
Support send msgs to exchange and flush msgs from exchange to parquet

### PR DESCRIPTION
**Pull request by wanghaEMQ**
1. Support sends msgs to exchange in parallel
2. Support flush msgs got from exchange to parquet.
3.

### CODE REVIEW

File: nanomq/apps/broker.c [hunk 1]
```diff
@@ -945,18 +945,17 @@ broker(conf *nanomq_conf)
...
+			nng_socket_set_ptr(*node->sock, NNG_OPT_EXCHANGE_BIND, ex);
 		}
 		log_debug("exchange %d init finished!\n", i);
 	}
 	// Webhook service
 	if (nanomq_conf->web_hook.enable) {
 		log_debug("Webhook service initialization");
 		start_webhook_service(nanomq_conf);
+		if (nanomq_conf->exchange.count > 0)
+			hook_exchange_init(nanomq_conf);
```
Comment (JaylinYu) [inline nanomq/apps/broker.c:958]: ??? should not be here
num_ctx is changed in the following init section

Comment (JaylinYu) [reply nanomq/apps/broker.c:958]: and exchange is not related to webhook

File: nanomq/apps/broker.c [hunk 2]
```diff
@@ -1048,6 +1047,10 @@ broker(conf *nanomq_conf)
 			    PROTO_HTTP_SERVER, db, db_ret, nanomq_conf);
 		}
 	}
+	// create exchange senders
+	if (nanomq_conf->web_hook.enable && nanomq_conf->exchange.count > 0) {
+		hook_exchange_sender_init(nanomq_conf, works);
```
Comment (JaylinYu) [inline nanomq/apps/broker.c:1052]: be aware the num of ctx/work.

File: nanomq/pub_handler.c [hunk 3]
```diff
@@ -1593,6 +1593,7 @@ decode_pub_message(nano_work *work, uint8_t proto)
 		// payload
 		pub_packet->payload.len =
 		    (uint32_t) (msg_len - (size_t) used_pos);
+		nng_msg_set_payload_ptr(msg, msg_body + pos);
```
Comment (JaylinYu) [inline nanomq/pub_handler.c:1596]: cool

**Review by JaylinYu — CHANGES_REQUESTED** [review]

File: nanomq/webhook_post.c [hunk 4]
```diff
@@ -255,31 +258,39 @@ hook_entry(nano_work *work, uint8_t reason)
...
 	if (ex_conf->count > 0 && nng_msg_get_type(work->msg) == CMD_PUBLISH &&
 	    work->flag == CMD_PUBLISH) {
-		// dup msg for now, or reuse it?
+		// dup msg for now, TODO or reuse it?
 		nng_msg *msg;
 		nng_msg_alloc(&msg, 0);
 		nng_msg_header_append(msg, nng_msg_header(work->msg), nng_msg_header_len(work->msg));
 		nng_msg_append(msg, nng_msg_body(work->msg), nng_msg_len(work->msg));
+		nng_msg_set_payload_ptr(msg, nng_msg_body(msg) +
+		        (nng_msg_payload_ptr(work->msg) - (uint8_t *)nng_msg_body(work->msg)));
```
Comment (OdyWayne) [inline nanomq/webhook_post.c:267]: something wrong here

Comment (wanghaEMQ) [reply nanomq/webhook_post.c:267]: where?

Comment (wanghaEMQ) [reply nanomq/webhook_post.c]: Fixed!

File: nanomq/apps/broker.c (hunk 2 above)
Comment (wanghaEMQ) [reply nanomq/apps/broker.c]: Yes. I have fixed it

File: nanomq/apps/broker.c (hunk 1 above)
Comment (wanghaEMQ) [reply nanomq/apps/broker.c]: Now exchange is an independent service to start

File: nanomq/webhook_post.c [hunk 5]
```diff
@@ -255,31 +258,40 @@ hook_entry(nano_work *work, uint8_t reason)
...
+				nng_aio *aio = hook_conf->saios[work->ctx.id-1];
 				int     *nkey = nng_alloc(sizeof(int));
-				*nkey         = g_msg_index;
-				nng_aio_alloc(&aio, NULL, NULL);
+
+				nng_mtx_lock(hook_conf->ex_mtx);
+				*nkey = g_msg_index++;
+				nng_mtx_unlock(hook_conf->ex_mtx);
+
+				if (nng_aio_busy(aio))
```
Comment (JaylinYu) [inline nanomq/webhook_post.c:284]: no need to check busy

**Review by JaylinYu — APPROVED** [review]

**Pull request merged**